File: exercism/python/twelve-days/twelve_days.py
#https://exercism.io/my/solutions/27c44b1163494641ac0470f4777858b6


def recite(start_verse, end_verse):

    DATA = [('first', 'a Partridge in a Pear Tree'),
            ('second', 'two Turtle Doves'),
            ('third', 'three French Hens'),
            ('fourth', 'four Calling Birds'),
            ('fifth', 'five Gold Rings'),
            ('sixth', 'six Geese-a-Laying'),
            ('seventh', 'seven Swans-a-Swimming'),
            ('eighth', 'eight Maids-a-Milking'),
            ('ninth', 'nine Ladies Dancing'),
            ('tenth', 'ten Lords-a-Leaping'),
            ('eleventh', 'eleven Pipers Piping'),
            ('twelfth', 'twelve Drummers Drumming')]

    def presents(d):
        last = DATA[0][1] if d == 1 else f', and {DATA[0][1]}'
        rest = ', '.join(DATA[j-1][1] for j in range(d, 1, -1))
        return rest + last

    def verse(d):
        return f'On the {DATA[d-1][0]} day of Christmas my true love gave to me: {presents(d)}.'

    return [verse(i) for i in range(start_verse, end_verse + 1)]

File: exercism/go/twelve-days/twelve_days.go
// https://exercism.io/my/solutions/e09fd5018b2b4c0cbb59c442ba84de07

package twelve

import (
	"fmt"
	"strings"
)

var (
	presents = []string{
		"twelve Drummers Drumming, ",
		"eleven Pipers Piping, ",
		"ten Lords-a-Leaping, ",
		"nine Ladies Dancing, ",
		"eight Maids-a-Milking, ",
		"seven Swans-a-Swimming, ",
		"six Geese-a-Laying, ",
		"five Gold Rings, ",
		"four Calling Birds, ",
		"three French Hens, ",
		"two Turtle Doves, and ",
		"a Partridge in a Pear Tree.",
	}
	ordinals = []string{
		"first",
		"second",
		"third",
		"fourth",
		"fifth",
		"sixth",
		"seventh",
		"eighth",
		"ninth",
		"tenth",
		"eleventh",
		"twelfth",
	}
	intro  = "On the %s day of Christmas my true love gave to me: "
	verses = make([]string, 12)
	song   string
)

// Generates every verse and the whole song only once
func init() {
	for i := 1; i <= 12; i++ {
		verses[i-1] = fmt.Sprintf(intro, ordinals[i-1]) + strings.Join(presents[12-i:], "")
	}
	song = strings.Join(verses, "\n")
}

// Verse returns the n-th verse of the Christmas song.
func Verse(n int) string {
	return verses[n-1]
}

// Song returns the (whole) Christmas song.
func Song() string {
	return song
}
